src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/HttpContentJsonExtensions.cs

```diff
@@ -0,0 +1,113 @@
+
+            // Wrap content stream into a transcoding stream that buffers the data transcoded from the sourceEncoding to utf-8.
+            if (sourceEncoding != null && sourceEncoding != Encoding.UTF8)
+            {
+                contentStream = new TranscodingReadStream(contentStream, sourceEncoding);
+            }
+
+            using (contentStream)
+            {
+                return await JsonSerializer.DeserializeAsync(contentStream, type, options, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private static async Task<T> ReadFromJsonAsyncCore<T>(HttpContent content, Encoding? sourceEncoding, JsonSerializerOptions? options, CancellationToken cancellationToken)
+        {
+            Stream contentStream = await content.ReadAsStreamAsync().ConfigureAwait(false);
+
+            // Wrap content stream into a transcoding stream that buffers the data transcoded from the sourceEncoding to utf-8.
+            if (sourceEncoding != null && sourceEncoding != Encoding.UTF8)
+            {
+                contentStream = new TranscodingReadStream(contentStream, sourceEncoding);
+            }
+
+            using (contentStream)
+            {
+                return await JsonSerializer.DeserializeAsync<T>(contentStream, options, cancellationToken).ConfigureAwait(false);
+            }
+        }
+
+        private static void ValidateContent(HttpContent content)
+        {
+            if (content == null)
+            {
+                throw new ArgumentNullException(nameof(content));
+            }
+
+            string? mediaType = content.Headers.ContentType?.MediaType;
+
+            if (mediaType == null ||
+                !mediaType.Equals(JsonContent.JsonMediaType, StringComparison.OrdinalIgnoreCase) &&
+                !IsValidStructuredSyntaxJsonSuffix(mediaType.AsSpan()))
+            {
+                throw new NotSupportedException(SR.ContentTypeNotSupported);
+            }
+        }
+
+        private static bool IsValidStructuredSyntaxJsonSuffix(ReadOnlySpan<char> mediaType)
+        {
+            int index = 0;
+            int typeLength = mediaType.IndexOf('/');
+
+            if (typeLength < 0 ||
+                mediaType.Slice(index, typeLength).CompareTo(JsonContent.JsonType.AsSpan(), StringComparison.OrdinalIgnoreCase) != 0)
+            {
+                return false;
+            }
+
+            index += typeLength + 1;
+            int suffixStart = mediaType.Slice(index).IndexOf('+');
+
+            // Empty prefix subtype ("application/+json") not allowed.
+            if (suffixStart <= 0)
+            {
+                return false;
+            }
+
+            index += suffixStart + 1;
+            ReadOnlySpan<char> mediaTypeSuffix = mediaType.Slice(index);
+            if (mediaTypeSuffix.CompareTo(JsonContent.JsonSubtype.AsSpan(), StringComparison.OrdinalIgnoreCase) != 0)
```

Comment:
Nit: in other cases above where you sliced, you didn't store it into a tmp local, but here you did.

Reply:
I thought doing just `Slice` in that condition was not clear enough; I can add locals to the rest of cases.